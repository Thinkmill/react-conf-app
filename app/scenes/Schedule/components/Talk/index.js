// @flow
import React, { Component } from 'react';
import {
  Animated,
  Easing,
  PixelRatio,
  StyleSheet,
  Text,
  TouchableHighlight,
  View,
} from 'react-native';
import Icon from 'react-native-vector-icons/Ionicons';

import Avatar from '../../../../components/Avatar';
import theme from '../../../../theme';
import { lighten } from '../../../../utils/color';
import type { Speaker as SpeakerType } from '../../../../types';

type Status = 'past' | 'present' | 'future';

// ==============================
// TALK STATUSBAR
// ==============================

export function TalkStatusBar({ status, ...props }: { status: Status }) {
  let barColor = theme.color.gray20;
  if (status === 'past') barColor = lighten(theme.color.blue, 60);
  if (status === 'present') barColor = theme.color.blue;

  return (
    <View
      style={{
        backgroundColor: barColor,
        width: 5,
      }}
      {...props}
    />
  );
}

// ==============================
// ICON HELPERS
// ==============================

function Indicator({ color, icon }) {
  return (
    <View
      style={{
        alignItems: 'center',
        backgroundColor: color,
        borderRadius: 14,
        marginRight: 7,
        height: 14,
        justifyContent: 'center',
        width: 14,
      }}
    >
      <Icon
        color="white"
        name={icon}
        size={14}
        style={{ backgroundColor: 'transparent', marginBottom: -1 }}
      />
    </View>
  );
}
function LightningSubtitle({ text, ...props }) {
  return (
    <View style={styles.subtitle} {...props}>
      <Indicator color={theme.color.yellow} icon="ios-flash" />
      <Text style={styles.subtitleText}>{text}</Text>
    </View>
  );
}
function KeynoteSubtitle({ text, ...props }) {
  return (
    <View style={styles.subtitle} {...props}>
      <Indicator color={theme.color.blue} icon="ios-key" />
      <Text style={styles.subtitleText}>{text}</Text>
    </View>
  );
}

// ==============================
// TALK ROW
// ==============================

type Props = {
  keynote?: boolean,
  lightning?: boolean,
  onPress: () => mixed,
  speakers: Array<SpeakerType>,
  startTime: string,
  status: Status,
  title: string,
};

const animationDefault = val => ({
  toValue: val,
  duration: 666,
  easing: Easing.inOut(Easing.quad),
});

export default class Talk extends Component {
  props: Props;
  animValue: Animated.Value;

  static defaultProps = {
    status: 'future',
  };

  constructor(props: Props) {
    super(props);

    this.animValue = new Animated.Value(0);
  }
  componentDidMount() {
    this.cycleAnimation();
  }
  cycleAnimation() {
    Animated.sequence([
      Animated.timing(this.animValue, animationDefault(1)),
      Animated.timing(this.animValue, animationDefault(0)),
    ]).start(() => this.cycleAnimation());
  }
  render() {
    const {
      keynote,
      lightning,
      onPress,
      speakers,
      startTime,
      status,
      title,
      ...props
    } = this.props;

    const isPresent = status === 'present';

    const touchableProps = {
      activeOpacity: 1,
      onPress: onPress,
      style: styles.touchable,
      underlayColor: theme.color.gray05,
    };
    const animatedStyle = {
      transform: [
        {
          translateX: this.animValue.interpolate({
            inputRange: [0, 1],
            outputRange: [0, 4],
          }),
        },
      ],
    };

    // subtitle variants
    let subtitleText = startTime;
    let speakersText;

    if (speakers) {
      speakersText = speakers.map(speaker => speaker.name).join(', ');
      subtitleText += ' - ' + speakersText;
    }

    let subtitle;

    if (lightning) {
      subtitle = <LightningSubtitle text={speakersText} />;
    } else if (keynote) {
      subtitle = <KeynoteSubtitle text={startTime} />;
    } else {
      subtitle = (
        <Text style={[styles.subtitle, styles.subtitleText]}>
          {subtitleText}
        </Text>
      );
    }

    // avatar variants
    const avatar = speakers
      ? speakers.map((speaker, index) => {
          const pull = index + 1 !== speakers.length
            ? { marginRight: -16 }
            : null;
          return (
            <Avatar key={speaker.name} source={speaker.avatar} style={pull} />
          );
        })
      : null;

    // const avatar = <Avatar source={speaker.avatar} />;

    return (
      <TouchableHighlight {...touchableProps} {...props}>
        <View style={[styles.base, styles['base__' + status]]}>
          <TalkStatusBar status={status}>
            {isPresent &&
              <Animated.View style={animatedStyle}>
                <Icon
                  color={theme.color.blue}
                  name="md-arrow-dropright"
                  size={34}
                  style={styles.statusbarIcon}
                />
              </Animated.View>}
          </TalkStatusBar>

          <View style={[styles.content, styles['content__' + status]]}>
            <View style={[styles.text, styles['text__' + status]]}>
              {subtitle}
              <Text style={[styles.title, styles['title__' + status]]}>
                {title}
              </Text>
            </View>

            <View style={styles.right}>
              {avatar}
              <Icon
                color={theme.color.gray40}
                name="ios-arrow-forward"
                size={20}
                style={styles.chevron}
              />
            </View>
          </View>
        </View>
      </TouchableHighlight>
    );
  }
}

const styles = StyleSheet.create({
  touchable: {
    backgroundColor: 'white',
  },
  base: {
    alignItems: 'stretch',
    backgroundColor: 'transparent',
    borderBottomColor: theme.color.gray20,
    borderBottomWidth: 1 / PixelRatio.get(),
    flexDirection: 'row',
  },
  // base__present: {
  // 	backgroundColor: fade(theme.color.blue, 3),
  // },

  statusbarIcon: {
    backgroundColor: 'transparent',
    height: 34,
    left: 0,
    position: 'absolute',
    top: 10,
    width: 34,
  },

  // content
  content: {
    alignItems: 'center',
    backgroundColor: 'transparent',
    flexDirection: 'row',
    flexGrow: 1,
    flexShrink: 1,
    padding: theme.fontSize.default,
  },
  content__past: {
    opacity: 0.5,
  },
  text: {
    flexGrow: 1,
    flexShrink: 1,
    paddingRight: theme.fontSize.xsmall,
  },
  subtitle: {
    alignItems: 'center',
    flexDirection: 'row',
    marginBottom: theme.fontSize.small,
  },
  subtitleText: {
    color: theme.color.gray60,
    flexShrink: 1,
    fontSize: theme.fontSize.small,
    fontWeight: '300',
  },
  title: {
    color: theme.color.text,
    fontSize: theme.fontSize.default,
  },

  // right (avatar and chevron)
  right: {
    alignItems: 'center',
    flexDirection: 'row',
    flexShrink: 0,
  },

  // chevron
  chevron: {
    marginLeft: theme.fontSize.default,
  },
});
