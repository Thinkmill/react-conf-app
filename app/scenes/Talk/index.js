// @flow
import React, { Component } from 'react';
import { Animated, Dimensions, Share } from 'react-native';
import moment from 'moment';

import type { ScheduleTalk } from '../../types';
import { TIME_FORMAT } from '../../constants';
import Navbar from '../../components/Navbar';
import Scene from '../../components/Scene';

import theme from '../../theme';
import { getNextTalkFromId, getPrevTalkFromId } from '../../data/talks';

import Hint from './components/Hint';
import Speaker from './components/Speaker';
import TalkPane from './components/Pane';

type Props = {
  navigator: Object,
  nextTalk: ScheduleTalk | null,
  prevTalk: ScheduleTalk | null,
  talk: ScheduleTalk,
  introduceUI: boolean,
};

type TransitionDirection = 'prev' | 'next';

type State = {
  animValue: Animated.Value,
  modalIsOpen: boolean,
  nextTalk: ScheduleTalk | null,
  prevTalk: ScheduleTalk | null,
  showIntro: boolean,
  talk: ScheduleTalk,
  incomingTalk?: ScheduleTalk | null,
  transitionDirection?: TransitionDirection,
};

type SetTalksState = {
  talk: ScheduleTalk,
  nextTalk: ScheduleTalk | null,
  prevTalk: ScheduleTalk | null,
};

export default class Talk extends Component {
  talkpane: $FlowFixMe; // https://github.com/facebook/flow/issues/2202
  transitionpane: $FlowFixMe; // https://github.com/facebook/flow/issues/2202

  props: Props;
  state: State = {
    animValue: new Animated.Value(0),
    modalIsOpen: false,
    nextTalk: this.props.nextTalk,
    prevTalk: this.props.prevTalk,
    showIntro: this.props.introduceUI,
    talk: this.props.talk,
  };

  sceneHeight = Dimensions.get('window').height;
  sceneWidth = Dimensions.get('window').width;

  handleLayout({ height }: { height: number }) {
    const availableHeight = this.sceneHeight - height;

    this.talkpane.refs.summary.measure((left, top, width, height) => {
      if (availableHeight > height) {
        this.talkpane.refs.summary.setNativeProps({
          style: { minHeight: availableHeight - theme.nextup.height },
        });
      }
    });
  }
  handleScroll = ({ nativeEvent }: Object) => {
    const contentHeight = nativeEvent.contentSize.height;
    const viewHeight = nativeEvent.layoutMeasurement.height;
    const scrollY = nativeEvent.contentOffset.y;
    const heightOffset = contentHeight > viewHeight
      ? contentHeight - viewHeight
      : 0;
    const { nextTalkPreview, prevTalkPreview, scrollview } = this.talkpane.refs;
    const { nextTalk, prevTalk } = this.state;

    if (scrollY > 0 && nextTalkPreview) {
      const opacity = Math.min((scrollY - (heightOffset + 20)) / 100, 1);
      nextTalkPreview.setNativeProps({ style: { opacity } });
    } else if (prevTalkPreview) {
      const opacity = Math.min(Math.abs(scrollY + 20) / 100, 1);
      prevTalkPreview.setNativeProps({ style: { opacity } });
    }

    const jumpToNext = nextTalk && scrollY > heightOffset + 110;
    const jumpToPrev = prevTalk && scrollY < -110;

    if (jumpToNext) {
      scrollview.setNativeProps({
        scrollEnabled: false,
        bounces: false,
      });
      this.renderNextTalk();
    } else if (jumpToPrev) {
      scrollview.setNativeProps({
        bounces: false,
        scrollEnabled: false,
      });
      this.renderPrevTalk();
    }
  };
  renderNextTalk() {
    const talk = this.state.nextTalk;
    const nextTalk = this.state.nextTalk
      ? getNextTalkFromId(this.state.nextTalk.id)
      : null;
    const prevTalk = this.state.talk;

    if (talk !== null) {
      this.setTalks({ nextTalk, prevTalk, talk }, 'next');
    }
  }
  renderPrevTalk() {
    const talk = this.state.prevTalk;
    const nextTalk = this.state.talk;
    const prevTalk = this.state.prevTalk
      ? getPrevTalkFromId(this.state.prevTalk.id)
      : null;

    if (talk !== null) {
      this.setTalks({ nextTalk, prevTalk, talk }, 'prev');
    }
  }
  setTalks(newState: SetTalksState, transitionDirection: TransitionDirection) {
    this.setState(
      {
        incomingTalk: newState.talk,
        transitionDirection,
      },
      () => {
        Animated.spring(this.state.animValue, {
          toValue: 1,
          friction: 7,
          tension: 30,
        }).start(() => {
          this.setState(
            Object.assign({}, newState, { incomingTalk: null }),
            () => {
              this.state.animValue.setValue(0);
              this.talkpane.refs.scrollview.setNativeProps({
                bounces: true,
                scrollEnabled: true,
              });
              this.talkpane.refs.scrollview.scrollTo({ y: 0, animated: false });
            }
          );
        });
      }
    );
  }
  share = () => {
    const { talk } = this.state;
    const speakerHandle = talk.speaker.twitter
      ? '@' + talk.speaker.twitter
      : talk.speaker.name;

    Share.share({
      title: 'ReactConf 2017',
      message: `Loving ${speakerHandle}'s talk "${talk.title}" #ReactConf2017`,
    });
  };
  toggleSpeakerModal = (data: Object) => {
    this.setState({
      modalIsOpen: !this.state.modalIsOpen,
      modalSpeaker: data,
    });
  };
  render() {
    const { navigator } = this.props;
    const {
      animValue,
      modalIsOpen,
      modalSpeaker,
      nextTalk,
      incomingTalk,
      prevTalk,
      showIntro,
      talk,
    } = this.state;

    const headerTitle = moment(talk.time.start).format(TIME_FORMAT);
    const availableHeight = this.sceneHeight - theme.navbar.height;

    const incomingFrom = this.state.transitionDirection === 'next'
      ? this.sceneHeight
      : -this.sceneHeight;
    const outgoingTo = this.state.transitionDirection === 'next'
      ? -this.sceneHeight
      : this.sceneHeight;

    const transitionStyles = {
      height: availableHeight,
      position: 'absolute',
      top: theme.navbar.height,
      width: this.sceneWidth,
    };
    const incomingTransitionStyles = {
      transform: [
        {
          translateY: animValue.interpolate({
            inputRange: [0, 1],
            outputRange: [incomingFrom, 0],
          }),
        },
      ],
    };
    const outgoingTransitionStyles = {
      transform: [
        {
          translateY: animValue.interpolate({
            inputRange: [0, 1],
            outputRange: [0, outgoingTo],
          }),
        },
      ],
    };

    // navbar must be rendered after the talk panes for visibility
    const navbar = (
      <Navbar
        title={headerTitle}
        leftButtonIconName="ios-arrow-back"
        leftButtonOnPress={navigator.popToTop}
        rightButtonText="Share"
        rightButtonOnPress={this.share}
      />
    );

    return (
      <Scene>
        <Animated.View style={[transitionStyles, outgoingTransitionStyles]}>
          <TalkPane
            nextTalk={nextTalk}
            onHeroLayout={({ nativeEvent: { layout } }) =>
              this.handleLayout(layout)}
            onScroll={this.handleScroll}
            prevTalk={prevTalk}
            ref={r => this.talkpane = r}
            showSpeakerModal={this.toggleSpeakerModal}
            visibleTalk={talk}
          />
        </Animated.View>
        {!!incomingTalk &&
          <Animated.View
            style={[transitionStyles, incomingTransitionStyles]}
            pointerEvents="none"
          >
            <TalkPane
              visibleTalk={incomingTalk}
              ref={r => this.transitionpane = r}
            />
          </Animated.View>}

        {showIntro &&
          <Hint onClose={() => this.setState({ showIntro: false })} />}

        {modalIsOpen &&
          <Speaker
            avatar={modalSpeaker.avatar}
            github={modalSpeaker.github}
            name={modalSpeaker.name}
            onClose={this.toggleSpeakerModal}
            summary={modalSpeaker.summary}
            twitter={modalSpeaker.twitter}
          />}
        {navbar}
      </Scene>
    );
  }
}
