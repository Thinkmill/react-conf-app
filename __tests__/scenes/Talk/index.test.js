// @flow
import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';

import Talk from '../../../app/scenes/Talk';

const talk = {
  id: '',
  summary: '',
  title: '',
  speaker: {
    avatar: '',
    github: '',
    name: '',
    twitter: '',
    summary: '',
  },
  time: {
    // These dates, contrary to their appearance, will end up at the same time
    // each time because we're using mockdate in the setup.js file.
    start: new Date(),
    end: new Date(),
  },
};

describe('Talk', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(
        <Talk
          navigator={{}}
          talk={talk}
          nextTalk={null}
          prevTalk={null}
          introduceUI={false}
        />
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
