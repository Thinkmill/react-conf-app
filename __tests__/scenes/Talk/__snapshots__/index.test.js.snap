exports[`Talk renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }>
  <View
    style={
      Object {
        "height": 1270,
        "position": "absolute",
        "top": 64,
        "transform": Array [
          Object {
            "translateY": 0,
          },
        ],
        "width": 750,
      }
    }>
    <RCTScrollView
      onScroll={[Function]}
      scrollEventThrottle={90}
      style={
        Object {
          "flex": 1,
        }
      }>
      <View>
        <View
          onLayout={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "white",
              "borderBottomColor": "#CCC",
              "borderBottomWidth": 0.5,
              "borderTopColor": "#CCC",
              "borderTopWidth": 0.5,
              "marginTop": -0.5,
              "paddingBottom": 32,
              "paddingHorizontal": 24,
              "paddingTop": 32,
            }
          }>
          <View
            accessibilityComponentType={undefined}
            accessibilityLabel={undefined}
            accessibilityTraits={undefined}
            accessible={true}
            hitSlop={undefined}
            isTVSelectable={true}
            onLayout={undefined}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
            testID={undefined}
            tvParallaxProperties={undefined}>
            <View
              style={
                Object {
                  "alignItems": "center",
                  "paddingHorizontal": 32,
                }
              }>
              <View
                style={
                  Array [
                    Object {
                      "backgroundColor": "#F4F4F4",
                      "borderRadius": 44,
                      "height": 44,
                      "overflow": "hidden",
                      "width": 44,
                    },
                    undefined,
                  ]
                }>
                <Image
                  source={
                    Object {
                      "uri": "",
                    }
                  }
                  style={
                    Object {
                      "borderRadius": 0,
                      "height": 44,
                      "width": 44,
                    }
                  } />
              </View>
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                style={
                  Object {
                    "color": "#00A8D8",
                    "fontSize": 17,
                    "fontWeight": "500",
                    "marginTop": 14,
                  }
                }>
                
              </Text>
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                style={
                  Object {
                    "color": "#999",
                    "fontSize": 12,
                    "paddingBottom": 24,
                  }
                }>
                (tap for more)
              </Text>
            </View>
          </View>
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "fontSize": 24,
                "fontWeight": "300",
                "textAlign": "center",
              }
            }>
            
          </Text>
        </View>
        <View
          style={
            Object {
              "paddingBottom": 40,
            }
          }>
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "fontSize": 17,
                "fontWeight": "300",
                "lineHeight": 24,
                "padding": 24,
              }
            }>
            
          </Text>
        </View>
      </View>
    </RCTScrollView>
  </View>
  <View>
    <View
      style={
        Object {
          "height": 64,
        }
      } />
    <View
      style={
        Array [
          Object {
            "alignItems": "stretch",
            "borderBottomColor": "#CCC",
            "borderBottomWidth": 0.5,
            "flexDirection": "row",
            "height": 64,
            "justifyContent": "space-between",
            "overflow": "hidden",
            "paddingTop": 20,
            "position": "absolute",
            "width": 750,
          },
          Object {
            "backgroundColor": "white",
          },
        ]
      }>
      <View
        style={
          Object {
            "alignItems": "center",
            "flex": 3,
            "flexDirection": "row",
            "paddingHorizontal": 17,
          }
        } />
      <View
        style={
          Object {
            "alignItems": "center",
            "flex": 4,
            "justifyContent": "center",
          }
        }>
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": "white",
                "fontSize": 17,
                "fontWeight": "500",
              },
              Object {
                "color": "#383838",
              },
            ]
          }>
          8:00am
        </Text>
      </View>
      <View
        accessibilityComponentType={undefined}
        accessibilityLabel={undefined}
        accessibilityTraits={undefined}
        accessible={true}
        hitSlop={undefined}
        isTVSelectable={true}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "flex": 3,
            "flexDirection": "row",
            "justifyContent": "flex-end",
            "opacity": 1,
            "paddingHorizontal": 17,
          }
        }
        testID={undefined}
        tvParallaxProperties={undefined}>
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": "#00A8D8",
                "fontSize": 17,
              },
              Object {
                "opacity": 1,
              },
            ]
          }>
          Share
        </Text>
      </View>
    </View>
  </View>
</View>
`;
